File: src/client/java/com/boclips/users/client/model/accessrule/AccessRule.java
package com.boclips.users.client.model.accessrule;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import lombok.Data;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
@JsonSubTypes({
        @JsonSubTypes.Type(value = IncludedCollectionsAccessRule.class, name = "IncludedCollections"),
        @JsonSubTypes.Type(value = IncludedVideosAccessRule.class, name = "IncludedVideos"),
        @JsonSubTypes.Type(value = ExcludedVideosAccessRule.class, name = "ExcludedVideos"),
        @JsonSubTypes.Type(value = ExcludedVideoTypesAccessRule.class, name = "ExcludedVideoTypes")
})
@Data
public abstract class AccessRule {
    protected String name;
}

File: src/main/kotlin/com/boclips/users/domain/model/contentpackage/AccessRule.kt
package com.boclips.users.domain.model.contentpackage

sealed class AccessRule {
    data class IncludedCollections(
        override val id: AccessRuleId,
        override val name: String,
        val collectionIds: List<CollectionId>
    ) : AccessRule()

    data class IncludedVideos(
        override val id: AccessRuleId,
        override val name: String,
        val videoIds: List<VideoId>
    ) : AccessRule()

    data class IncludedDistributionMethods(
        override val id: AccessRuleId,
        override val name: String,
        val distributionMethods: Set<DistributionMethod>
    ) : AccessRule()

    data class ExcludedVideos(
        override val id: AccessRuleId,
        override val name: String,
        val videoIds: List<VideoId>
    ) : AccessRule()

    data class ExcludedVideoTypes(
        override val id: AccessRuleId,
        override val name: String,
        val videoTypes: List<VideoType>
    ) : AccessRule()

    data class ExcludedContentPartners(
        override val id: AccessRuleId,
        override val name: String,
        val contentPartnerIds: List<ContentPartnerId>
    ) : AccessRule()

    abstract val id: AccessRuleId
    abstract val name: String
}
